src/schema/sources.ts

```diff
@@ -529,12 +553,36 @@ export const ensureSourcePermissions = async (
   validateRankAgainstId?: string,
 ): Promise<Source> => {
   if (sourceId) {
-    const source = await ctx.con
-      .getRepository(Source)
-      .findOneByOrFail([{ id: sourceId }, { handle: sourceId }]);
-    if (await canAccessSource(ctx, source, permission, validateRankAgainstId)) {
-      return source;
+    const [source, sourceMember] = await Promise.all([
+      ctx.con
+        .getRepository(Source)
+        .findOneByOrFail([{ id: sourceId }, { handle: sourceId }]),
+      ctx.con
+        .getRepository(SourceMember)
+        .findOneBy({ sourceId, userId: ctx.userId }),
+    ]);
+
+    const canAccess = await canAccessSource(
+      ctx,
+      source,
+      sourceMember,
+      permission,
+      validateRankAgainstId,
+    );
+
+    if (!canAccess) {
+      throw new ForbiddenError('Access denied!');
     }
+
+    if (
```

Comment:
I think if the source type is not a squad this condition should also throw an error right?